vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js

```diff
@@ -1,140 +1,146 @@
-
-            datepicker.i18n.formatDate = function (date) {
-                let rawDate = new Date(Date.UTC(date.year, date.month, date.day));
-                return cleanString(rawDate.toLocaleDateString(locale, { timeZone: 'UTC' }));
-            };
-
-            datepicker.i18n.parseDate = function (dateString) {
-                dateString = cleanString(dateString);
-
-                if (dateString.length == 0) {
-                    return;
+                let currentDate = false;
+                let inputValue = getInputValue();
+                if (datepicker.i18n.parseDate !== 'undefined' && inputValue) {
+                    /* get current date with old parsing */
+                    currentDate = datepicker.i18n.parseDate(inputValue);
                 }
 
-                let match = dateString.match(datepicker.$connector.regex);
-                if (match && match.length == 4) {
-                    for (let i = 1; i < 4; i++) {
-                        datepicker.$connector.parts[i-1].value = parseInt(match[i]);
+                /* create test-string where to extract parsing regex */
+                let testDate = new Date(Date.UTC(datepicker.$connector.yearPart.initial, datepicker.$connector.monthPart.initial - 1, datepicker.$connector.dayPart.initial));
+                let testString = cleanString(testDate.toLocaleDateString(locale, { timeZone: 'UTC' }));
+                datepicker.$connector.parts.forEach(function (part) {
+                    part.index = testString.indexOf(part.initial);
+                });
+                /* sort items to match correct places in regex groups */
+                datepicker.$connector.parts.sort(FlowDatePickerPart.compare);
+                /* create regex
+                * regex will be the date, so that:
+                * - day-part is '(\d{1,2})' (1 or 2 digits),
+                * - month-part is '(\d{1,2})' (1 or 2 digits),
+                * - year-part is '(\d{4})' (4 digits)
+                *
+                * and everything else is left as is.
+                * For example, us date "10/20/2010" => "(\d{1,2})/(\d{1,2})/(\d{4})".
+                *
+                * The sorting part solves that which part is which (for example,
+                * here the first part is month, second day and third year)
+                *  */
+                datepicker.$connector.regex = testString.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&').replace(datepicker.$connector.dayPart.initial, "(\\d{1,2})").replace(datepicker.$connector.monthPart.initial, "(\\d{1,2})").replace(datepicker.$connector.yearPart.initial, "(\\d{4})");
+
+                datepicker.i18n.formatDate = function (date) {
+                    let rawDate = new Date(Date.UTC(date.year, date.month, date.day));
+                    return cleanString(rawDate.toLocaleDateString(locale, { timeZone: 'UTC' }));
+                };
+
+                datepicker.i18n.parseDate = function (dateString) {
```

Comment:
I think these `formatDate` and `parseDate` methods should be wrapped with the `tryCatchWrapper` and then we don't need the wrapping directly on `cleanString` and `getInputValue` since those are only called within other wrapped methods then.

Reply:
I was trying to avoid direct wrapping of i18n functions with tryCatchWrapper and instead wrap the relative methods in order not to mess-up those with wrappers from other components on the page when debugging.

Updated now.